python/lsst/daf/butler/registry/tests/_registry.py

```diff
@@ -1668,3 +1668,81 @@ def testIngestTimeQuery(self):
         datasets = list(registry.queryDatasets(..., collections=..., where=where))
         len2 = len(datasets)
         self.assertEqual(len2, 0)
+
+    def testTimespanQueries(self):
+        """Test query expressions involving timespans.
+        """
+        registry = self.makeRegistry()
+        self.loadData(registry, "hsc-rc2-subset.yaml")
+        # All exposures in the database; mapping from ID to timespan.
+        visits = {record.id: record.timespan for record in registry.queryDimensionRecords("visit")}
+        # Just those IDs, sorted (which is also temporal sorting, because HSC
+        # exposure IDs are monotonically increasing).
+        ids = sorted(visits.keys())
+        self.assertGreater(len(ids), 20)
+        # Pick some quasi-random indexes into `ids` to play with.
+        i1 = int(len(ids)*0.1)
+        i2 = int(len(ids)*0.3)
+        i3 = int(len(ids)*0.6)
+        i4 = int(len(ids)*0.8)
+        # Extract some times from those: just before the beginning of i1 (which
+        # should be after the end of the exposure before), exactly the
+        # beginning of i2, just after the beginning of i3 (and before its end),
+        # and the exact end of i4.
+        t1 = visits[ids[i1]].begin - astropy.time.TimeDelta(1.0, format="sec")
+        self.assertGreater(t1, visits[ids[i1 - 1]].end)
+        t2 = visits[ids[i2]].begin
+        t3 = visits[ids[i3]].begin + astropy.time.TimeDelta(1.0, format="sec")
+        self.assertLess(t3, visits[ids[i3]].end)
+        t4 = visits[ids[i4]].end
+        # Make sure those are actually in order.
+        self.assertEqual([t1, t2, t3, t4], sorted([t4, t3, t2, t1]))
+
+        bind = {
+            "t1": t1,
+            "t2": t2,
+            "t3": t3,
+            "t4": t4,
+            "ts23": Timespan(t2, t3),
+        }
+
+        def query(where):
+            """Helper function that queries for visit data IDs and returns
+            results as a sorted, deduplicated list of visit IDs.
+            """
+            return sorted(
+                {dataId["visit"] for dataId in registry.queryDataIds("visit",
+                                                                     instrument="HSC",
+                                                                     bind=bind,
+                                                                     where=where)}
+            )
+
+        # Try a bunch of timespan queries, mixing up the bounds themselves,
+        # where they appear in the expression, and how we get the timespan into
+        # the expression.
+
+        # t1 is before the start of i1, so this should not include i1.
+        self.assertEqual(ids[:i1], query("visit.timespan OVERLAPS (None, t1)"))
```

Comment:
Hmm, I'm confused here, I guess "None" identifier is converted into Python `None` somehow but I don't remember seeing "None" handling in the code and it's not in the `bind`?

Reply:
I added it as a constant entry in the dict in a higher-level class, but since then I've realized I should move it (probably make a `categorizeConstant` function or something) in order to make it more obvious and make it case-insensitive.  I'll probably just support `null` instead of `null` and `None`.

Reply:
+1 for `null` only, too confusing to have two different constants meaning the same thing.